### PR TITLE
toga_cocoa: Implement the Window set_position method.

### PR DESCRIPTION
**Pull request by eliteraspberries**
The coordinate system on Cocoa has its origin at the bottom left corner of the screen, and the [setFrameOrigin](https://developer.apple.com/documentation/appkit/nswindow/1419690-setframeorigin?language=objc) method of NSWindow sets the position of the bottom-left corner of the window.

To be consistent with the other backends (position is from the top left corner of the screen to the top left corner of the window), I use [setFrameTopLeftPoint](https://developer.apple.com/documentation/appkit/nswindow/1419658-setframetopleftpoint?language=objc) and flip the y coordinate.

## PR Checklist:

- [x] All new features have been tested
- [ ] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

**Review by samschott — CHANGES_REQUESTED** [review]
This already looks good, many thanks! There are however a few corner cases which you have not considered:

1. The current code raises an error if the window is not currently on any screen.
2. The window may end up in the wrong position for multiple screen setups.

Have a look at the `Window.create()` method to see how multiple screens can be handled. Toga currently takes the top-left corner of the main window as its coordinate system origin where the main window is the window with keyboard focus.

**Comment by eliteraspberries** [comment]
Ah yes, I didn't try with multiple screens. I'll do that.

File: src/cocoa/toga_cocoa/window.py [hunk 1]
```diff
@@ -222,7 +223,18 @@ def set_title(self, title):
...
+        screen = self.native.screen
+        frame = screen.frame
+
+        # From the bottom left corner of the screen to the bottom left corner
+        # of the window.
+        # self.native.setFrameOrigin(NSPoint(x, y))
+
+        # From the top left corner of the screen to the top left corner
+        # of the window.
+        self.native.setFrameTopLeftPoint(NSPoint(x, frame.size.height - y))
```
[the remark this answers is not in the document]
Comment (freakboy3742) [reply src/cocoa/toga_cocoa/window.py:238]: I wasn't able to find any cases where this would *break* - did you have any specific cases in mind @samschott?

However, there's definitely some problems where y < (height of the title bar). Setting position to (0, 0) will position you at the origin (just below the menubar)... but so will (0, 10) and (0, 20). Do we need to account for `NSMenu.menuBarHeight` and/or `NSMenu.menuBarVisible?`? Or should we just accept the fact that (0, 0) is a bit of a "false" origin?

File: src/cocoa/toga_cocoa/window.py [hunk 2]
```diff
@@ -222,7 +223,20 @@ def set_title(self, title):
...
+
+        x, y = position
+        screen = self.native.screen
+        if screen is None:
+            return
+        frame = screen.frame
+
+        # From the bottom left corner of the screen to the bottom left corner
+        # of the window.
+        # self.native.setFrameOrigin(NSPoint(x, y))
```
Comment (freakboy3742) [inline src/cocoa/toga_cocoa/window.py]: We don't need to retain old code like this; it can be deleted if not required.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
Thanks for the patch! I've made a couple of small tweaks; most importantly, the addition of an example. There's one open question about how to handle the menubar, combined with Sam's question about problem edge cases.

As a request for future patches - having a test case is an important part of every patch, because it makes it possible for reviewers to see what you've used to validate the code. Ideally, this will be a modification to an existing example (as I've done here), adding some small demonstration of the new/fixed feature. 

Alternatively, you can add an entirely new example. This is generally the preferred option if if you're adding an entirely new widget, or a completely new feature. We don't want an endless proliferation of examples, so if we can add a feature to an existing example, that's the preferred option.

The last option is to provide a standalone test app - a minimal app in a gist or standalone repo that demonstrates the functionality you've added. This is the approach to take if you're fixing a complex bug that is difficult to add to an existing example.

File: src/cocoa/toga_cocoa/window.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (samschott) [reply src/cocoa/toga_cocoa/window.py:238]: The problem arises mostly when having two screens with different resolutions. Have a look at this sketch as an illustration:

IMG_0654

Regarding the menu bar, I would just accept that (0,0) is a false origin. In particular, when the menu bar is hidden, macOS will probably correctly position the window at the top again.

File: src/cocoa/toga_cocoa/window.py [hunk 3]
```diff
@@ -222,7 +223,19 @@ def set_title(self, title):
...
 
     def set_position(self, position):
-        pass
+        x, y = position
+        screen = self.native.screen
+
+        # If the window isn't visible, it won't be on a screen;
+        # don't try to set the position.
+        if screen is None:
+            return
```
Comment (samschott) [inline src/cocoa/toga_cocoa/window.py:232]: I’m not sure if we should just give up here. I haven‘t tested it yet but ˋscreenˋ will probably also be ˋNoneˋ if the window was previously dragged or positioned off–screen instead of being set to not visible.

it might be better to then take the primary screen as a reference.

File: src/cocoa/toga_cocoa/window.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (samschott) [reply src/cocoa/toga_cocoa/window.py:238]: @freakboy3742, there was a good reason why your initial code ended but being a bit more complicated — it wasn’t just to confuse us ;)

[the remark this answers is not in the document]
Comment (eliteraspberries) [reply src/cocoa/toga_cocoa/window.py:238]: That illustration helps a lot. So we need the coordinates of this screen's origin relative to the "main" screen. Now I get it.

File: src/cocoa/toga_cocoa/window.py (hunk 3 above)
Comment (eliteraspberries) [reply src/cocoa/toga_cocoa/window.py:232]: I guess we fall back to the main screen in that case? The main screen is the one with the keyboard focus.

**Comment by eliteraspberries** [comment]
Sorry for the delay, I'm just waiting for a miniDP to HDMI adapter. My understanding of the screen's relative origin may be way off.

File: src/cocoa/toga_cocoa/window.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (samschott) [reply src/cocoa/toga_cocoa/window.py]: Yes, exactly :)

**Comment by samschott** [comment]
While we are discussing position, something which has been bothering me a little bit: The `create()` implementation handles positioning by treating the (x, y) coordinates always relative to the screen which holds keyboard focus. @eliteraspberries, your `setPosition()` implementation treats the position as relative to the screen where the window is currently on.

1. We should unify this bahavior.
2. Either option does not allow moving the window between screens by using the position property. This seems like a significant limitation.

Would it make sense to instead treat the position always relative to the top-left corder of the "primary" screen instead? We could you `NSScreen.screens` for that:

> The screen at index 0 in the returned array corresponds to the primary screen of the user’s system. This is the screen that contains the menu bar and whose origin is at the point (0, 0).

File: src/cocoa/toga_cocoa/window.py (hunk 3 above)
Comment (samschott) [reply src/cocoa/toga_cocoa/window.py:232]: That could make sense. See my comment below on the current ambiguity about which screen to use.

**Comment by eliteraspberries** [comment]
Hmm... you're right. Mainly I wanted window positioning so that an app can save its position/size on closing and re-open in the same spot. If it could re-open on the same screen too, that would be nice.

So I think positioning relative to the primary screen may be the best solution, but it's tricky because screens can come and go. In that case we need to check if the requested position is "out of bounds," and fall back to the main screen, or the nearest one?

For example, right now the "To the right screen" button in the example doesn't move the window off screen for me (one screen), it's pushed *mostly* off screen but the close button is still visible. And if I request too big a y-value, it doesn't go off the bottom, it goes just far enough that the title bar is still visible. I'll work on these cases.

**Comment by eliteraspberries** [comment]
I think we cover all edge cases now, at least as much as I can test.

File: src/cocoa/toga_cocoa/window.py [hunk 4]
```diff
@@ -221,13 +222,50 @@ def set_content(self, widget):
...
+        self.native.setFrameTopLeftPoint(NSPoint(x, y))
+
+        # Update the actual position.
+        self.update_position()
+
+        # If we ended up off screen, or not where requested,
+        # fall back to the "main" screen.
+        if screen is None or not self.interface._position == position:
+            frame = NSScreen.mainScreen.frame
+            self.set_position((frame.origin.x + 100, frame.origin.y + 100))
```
Comment (samschott) [inline src/cocoa/toga_cocoa/window.py:262]: Why do you choose the (100, 100) relative to "main screen" as a fallback instead of just accepting the best effort of macOS?

For example, when calling `setPosition((0,0))`, macOS would position the window in the top-left corner and possible move it down by 25 pixels to accommodate the menu bar. This seems like perfectly reasonable behaviour to me and more in line with expectations than moving the window to (100, 100).

File: src/cocoa/toga_cocoa/window.py [hunk 5]
```diff
@@ -221,13 +222,50 @@ def set_content(self, widget):
...
+        x = frame.origin.x + x
+        y = frame.origin.y + frame.size.height - y
+        self.native.setFrameTopLeftPoint(NSPoint(x, y))
+
+        # Update the actual position.
+        self.update_position()
+
+        # If we ended up off screen, or not where requested,
+        # fall back to the "main" screen.
+        if screen is None or not self.interface._position == position:
```
Comment (samschott) [inline src/cocoa/toga_cocoa/window.py:260]: See above comment regarding `screen is None`.

File: examples/button/button/app.py [hunk 6]
```diff
@@ -117,6 +140,32 @@ def callbackSmaller(self, button):
...
 
+    def seek_origin(self, button):
+        # Move the window to the somewhere near the top-left of the main screen
+        self.main_window.position = (100, 30)
+
+    def seek_right_screen(self, button):
+        # Move the window off the right hand side of the main screen.
+        # If the user has multiple monitors with a screen to the right
+        # of the main screen, the window should appear there.
+        self.main_window.position = (2000, 400)
```
Comment (samschott) [inline examples/button/button/app.py:151]: There is no guarantee that this actually gives us the right screen since the primary screen may have a width > 2000 pixels.

File: src/cocoa/toga_cocoa/window.py [hunk 7]
```diff
@@ -221,13 +222,50 @@ def set_content(self, widget):
     def set_title(self, title):
         self.native.title = title
 
+    def update_position(self):
```
Comment (samschott) [inline src/cocoa/toga_cocoa/window.py:225]: This probably should be a private method since it's only used internally.

**Review by samschott — CHANGES_REQUESTED** [review]
This already looks a lot better to me. I've left a few comments inline which should be easy to address.

Regarding the discussion of whether to interpret the position relative to the primary screen: I've checked with the other platform implementations and it seems like none of them currently support setting the position. We would therefore be setting a precedent here. IMO it makes sense to take the primary screen instead of the current screen or the one with keyboard focus as a reference. This makes the position a reproducible and reliable quantity.  @freakboy3742, let me know if you object.

There is one larger issue which I have discovered when trying the modified example app: After dragging the window manually, the "Bump left" and "Bump right" buttons no longer work as expected. This is because `toga.Window.position` does not get updated after dragging the window through the GUI. I can see two ways to solve this:

1. Implement a delegate method `windowDidMove:` to get notified on window moves and update the position every time.
2. Change the property getter for `toga.Window.position` to request the position from the implementation layer instead of caching it. This would require defining a `get_position` method on all backends (you don't need to implement it on other backends).

Both options would effectively do the same work as the current `update_position` method but would ensure that value of `toga.Window.position` is always correct.

Note to self: We currently have a similar issue with `toga.Window.size`.

File: src/cocoa/toga_cocoa/window.py (hunk 4 above)
Comment (eliteraspberries) [reply src/cocoa/toga_cocoa/window.py]: I was worried the window might disappear off screen, which was silly, AppKit takes care of that.

**Comment by eliteraspberries** [comment]
The test in the last commit is kind of funny because I don't actually see a window appear, but it passes.

File: examples/button/button/app.py (hunk 6 above)
Comment (freakboy3742) [reply examples/button/button/app.py:151]: Completely agreed. 2000 "works for me", but that's hardly reliable. Ideally, we'd be able to get reliable screen reference information as part of the Toga API... but this yak is already sporting a magnificent hairstyle... :-)

**Comment by freakboy3742** [comment]
> This already looks a lot better to me. I've left a few comments inline which should be easy to address.
> 
> Regarding the discussion of whether to interpret the position relative to the primary screen: I've checked with the other platform implementations and it seems like none of them currently support setting the position. We would therefore be setting a precedent here. IMO it makes sense to take the primary screen instead of the current screen or the one with keyboard focus as a reference. This makes the position a reproducible and reliable quantity. @freakboy3742, let me know if you object.

Making the (0,0) the origin on the "primary" screen makes sense to me. It's a reliable and consistent coordinate system; the "current" screen strikes me as something that will be (a) a harder API to use, because the values you want to use will be highly variable, and (b) harder to implement.
 
> There is one larger issue which I have discovered when trying the modified example app: After dragging the window manually, the "Bump left" and "Bump right" buttons no longer work as expected. This is because `toga.Window.position` does not get updated after dragging the window through the GUI. I can see two ways to solve this:
> 
> 1. Implement a delegate method `windowDidMove:` to get notified on window moves and update the position every time.
> 2. Change the property getter for `toga.Window.position` to request the position from the implementation layer instead of caching it. This would require defining a `get_position` method on all backends (you don't need to implement it on other backends).
> 
> Both options would effectively do the same work as the current `update_position` method but would ensure that value of `toga.Window.position` is always correct.

I'd say option 2 is the better approach here. As currently implemented, we've got 2 sources of truth for position (one on the interface, and one on the native layer). The native layer will be (by definition) canonical, so I'd lean to treating it as canonical, and making the interface API a direct query of that canonical source.

> Note to self: We currently have a similar issue with `toga.Window.size`.

Noted :-) ...and I think the same logic applies for size (and, for that matter, any other property that can be determined canonically from an API)

**Comment by freakboy3742** [comment]
> The test in the last commit is kind of funny because I don't actually see a window appear, but it passes.

That's not surprising - the test suite runs with a "Dummy" backend whose entire purpose is to implement the API without having any physical manifestation. This makes it a reliable (and inspectable) test harness, without relying on the specifics of a given platform implementation.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
Thanks for this contribution! There's an overlap between this PR and #1023; and both PRs are subject to the "source of truth" problem. I've added #1461 to merge the two PRs, adding fixes to address the source of truth issue. As a result, I'm going to close this PR; but it was really helpful in getting that PR working, so thanks for the assist!
